### PR TITLE
make the proxies order the same as the order in config file

### PR DESCRIPTION
**Pull request by ezksd**
the order of proxies and Selector's proxies list are arbitrary, when you have dozens of nodes, it was confusing, I make it the same as the order in the config file and the change will not break up previous versions of web version GUI(http://clash.razord.top). I will create a pull request to clash-dashboard latter.

### CODE REVIEW

**Comment by Dreamacro** [comment]
Your implementation increased complexity. why not just add an ordered array in proxy groups.

**Comment by ezksd** [comment]
> Your implementation increased complexity. why not just add an ordered array in proxy groups.

there are two parts of it and they are different,  if  add an ordered array to proxy group, it will still need an ordered array in the outside,  proxies and proxy group was in the same level in JSON response of `http://127.0.0.1:9090/proxies`

```json
{
     "proxies" : 
         [
             "GLOBAL" : {all,now,type,history ,new ordered array}, 
			 "Proxy" : {all,now,type,history ,new ordered array}, 
			 "direct,reject url test,load balance ... and others proxy group..",
             "proxy1","proxy2","proxy3" ....
         ]
     "ordered array use to save the order of proxy123": ....
}
```

The selector is the only one using a map to store proxies, it needs extra space to store order information in Selector sturct. and the `all` field of proxy group is already an ordered array of string,(but the clash-dashboard use Set to resort it), add a new one seems not appropriate.  an `index` field was useful, maybe the other type of proxy group need it in the future (such as ping of speed test result ).

Or add ordered array outside, remove the `index` field of proxy. So "complexity" means restful api complexity or complexity of clash ?

**Comment by Dreamacro** [comment]
> Or add ordered array outside, remove the index field of proxy. So "complexity" means restful api complexity or complexity of clash ?

The first goal is to keep clash data flow simple and maintainable. the index is not a part of the `Proxy` struct but for `Proxy Group`. So I think this implementation is unreasonable. BTW, the `GLOBAL` is a `selector` and it contains all proxy node.

> The selector is the only one using a map to store proxies, it needs extra space to store order information in Selector sturct. and the all field of proxy group is already an ordered array of string,(but the clash-dashboard use Set to resort it), add a new one seems not appropriate. an index field was useful, maybe the other type of proxy group need it in the future (such as ping of speed test result ).

I think the extra space is not a bad choice when we focus on maintainability. And there is no burden to store the order of proxies (no multiple changes). As for compatibility, Is there any way to keep the RESTful API response (or just extend)? if not, this feature needs to be compatible until the next major version.

**Comment by ezksd** [comment]
> > Or add ordered array outside, remove the index field of proxy. So "complexity" means restful api complexity or complexity of clash ?
> 
> The first goal is to keep clash data flow simple and maintainable. the index is not a part of the `Proxy` struct but for `Proxy Group`. So I think this implementation is unreasonable. BTW, the `GLOBAL` is a `selector` and it contains all proxy node.
> 
> > The selector is the only one using a map to store proxies, it needs extra space to store order information in Selector sturct. and the all field of proxy group is already an ordered array of string,(but the clash-dashboard use Set to resort it), add a new one seems not appropriate. an index field was useful, maybe the other type of proxy group need it in the future (such as ping of speed test result ).
> 
> I think the extra space is not a bad choice when we focus on maintainability. And there is no burden to store the order of proxies (no multiple changes). As for compatibility, Is there any way to keep the RESTful API response (or just extend)? if not, this feature needs to be compatible until the next major version.



> > Or add ordered array outside, remove the index field of proxy. So "complexity" means restful api complexity or complexity of clash ?
> 
> The first goal is to keep clash data flow simple and maintainable. the index is not a part of the `Proxy` struct but for `Proxy Group`. So I think this implementation is unreasonable. BTW, the `GLOBAL` is a `selector` and it contains all proxy node.
> 
> > The selector is the only one using a map to store proxies, it needs extra space to store order information in Selector sturct. and the all field of proxy group is already an ordered array of string,(but the clash-dashboard use Set to resort it), add a new one seems not appropriate. an index field was useful, maybe the other type of proxy group need it in the future (such as ping of speed test result ).
> 
> I think the extra space is not a bad choice when we focus on maintainability. And there is no burden to store the order of proxies (no multiple changes). As for compatibility, Is there any way to keep the RESTful API response (or just extend)? if not, this feature needs to be compatible until the next major version.

I add a `[]string` field in Selector, no other changes, and as you say using "GLOBAL" in clash-dashboard, please review it again.

File: adapters/outbound/selector.go [hunk 1]
```diff
@@ -64,8 +65,10 @@ func NewSelector(name string, proxies []C.Proxy) (*Selector, error) {
 	}
 
 	mapping := make(map[string]C.Proxy)
-	for _, proxy := range proxies {
+	proxyList := make([]string, len(proxies))
+	for id, proxy := range proxies {
```
Comment (Dreamacro) [inline adapters/outbound/selector.go]: `id` ==> `idx`

File: config/config.go [hunk 2]
```diff
@@ -194,13 +194,15 @@ func parseGeneral(cfg *rawConfig) (*General, error) {
...
 	proxies := make(map[string]C.Proxy)
+	proxyList := []string{};
 	proxiesConfig := cfg.Proxy
 	groupsConfig := cfg.ProxyGroup
 
 	decoder := structure.NewDecoder(structure.Option{TagName: "proxy", WeaklyTypedInput: true})
 
 	proxies["DIRECT"] = adapters.NewProxy(adapters.NewDirect())
 	proxies["REJECT"] = adapters.NewProxy(adapters.NewReject())
+	proxyList = append(proxyList, "DIRECT","REJECT")
```
Comment (beyondkmp) [inline config/config.go]: formatting your code.

**Comment by Dreamacro** [comment]
please run `gofmt -s -w .` in root clash directory and remove https://github.com/Dreamacro/clash/pull/180/files#diff-9839b3bfffd8216d43eb00601167ebfeR37 about `all`

**Comment by ezksd** [comment]
> please run `gofmt -s -w .` in root clash directory and remove https://github.com/Dreamacro/clash/pull/180/files#diff-9839b3bfffd8216d43eb00601167ebfeR37 about `all`

done

**Comment by ezksd** [comment]
> please run `gofmt -s -w .` in root clash directory and remove https://github.com/Dreamacro/clash/pull/180/files#diff-9839b3bfffd8216d43eb00601167ebfeR37 about `all`

done

**Pull request merged**